WordPress/gutenberg
# Block Templates: Check the validity of the block list against the defined template

Fixes #4476

When editing a post, the template is ignored entirely which means that if you update the template and opens an existing post in the editor, the template won't be updated.

This PR adds a mechanism to check the content of the current post against the template. The check is considered valid if all the block names of the content match the block names of the template.

If the check is invalid a notice is displayed inviting the user to reset its post according to the template (which could restult in some content loss) or keep the content as is.

If you edit the post in the text editor, the check is also performed.

**Note** This PR is necessary in order to achieve nested template locking because we want to match which container block has which lock config in the template.

<img width="1004" alt="screen shot 2018-02-20 at 13 07 26" src="https://user-images.githubusercontent.com/272444/36423143-0da568fa-163f-11e8-8b02-4c4db8df7b41.png">

**Testing instructions**

You can make use of the [GCF](https://wordpress.org/plugins/gutenberg-custom-fields/) plugin to test this quickly.

 - Add a locked template for your post
 - Create and save a post
 - Edit the post in the texteditor and remove some blocks
 - When clicking outside of the text editor, the notice is shown
 - When you refresh the page the notice is shown.
 - Try both buttons in the notice.
